### ISSUE TITLE
Unhandled endpoint: /session/.../timeouts

### ISSUE THREAD

**Issue by arompa** [appium/appium-xcuitest-driver#29]
Hello,

When trying to run my tests I get the following exception:
`org.openqa.selenium.WebDriverException: Unhandled endpoint: /session/6E4763B8-D534-4B5A-88CF-AE736FD31C93/timeouts -- http://localhost:8100/ with parameters {
    wildcards =     (
        "session/6E4763B8-D534-4B5A-88CF-AE736FD31C93/timeouts"
    );
}`

Because of that issue I'm unable to test the new driver and I'd love to check it out,
so if anyone knows a work-around for this please share it with me :)

**More info:**

- **Server logs:** [LINK](http://pasted.co/dbb5e511/fullscreen.php?hash=df10e6ad4e22b808049adf51df85443b&toolbar=true&linenum=true)
- **IDE Stacktrace:** [LINK](http://pasted.co/7d47de47/fullscreen.php?hash=9e9f0e8047fbdd8ff4d3874ef54b0347&toolbar=true&linenum=true)
- **Selenium-java  version:** 2.53.0
- **Java-client version:** 3.4.1

What I did:

1) I cloned the repo using: 
```
git clone --recursive https://github.com/appium/appium-xcuitest-driver.git
```
2) Built the project: 
```
rm -rf node-modules
npm install
gulp transpile
```
3) Started the server
```
node .
```
4) Run tests in my IDE

If some more information is required please let me know and I'll gladly provide that.

Thanks

**Comment by kaleposhobios** [comment]
Hi,

I think what you are seeing is expected. /timeouts is indeed not a supported endpoint. This driver proxies the requests to Facebook WebDriverAgent, which does the actual communication with the simulator/device. Looking at the Facebook code:
https://github.com/facebook/WebDriverAgent/blob/master/WebDriverAgentLib/Commands/FBCustomCommands.m

you can see that it's expecting `timeouts/implicit_wait.` But even more so, it looks like they intentionally don't support that command: ```
+ (id<FBResponsePayload>)handleImplicitWaitCommand:(FBRouteRequest *)request
{
  // This method is intentionally not supported.
  return FBResponseDictionaryWithOK();
}
```
I don't know if there's a bug in the Appium driver or not; the Selenium spec seems to indicate it ought to be ```timeouts/implicitlyWait```. But anyway it doesn't matter because this driver won't support that.

Can you just take it out and proceed without it?
Regards,
Matthew

**Comment by arompa** [comment]
Hi Matthew,

Thank you for the fast reply! :)

I am not using implicit waits in my test, but I think I know where the problem is.
I am using Page Factory and AppiumFieldDecorator which I believe uses the implicit wait..

I am wondering if anybody using the Facebook WebDriverAgent encountered this problem before?
Do you maybe have some proposition on what to replace the following with:
`        PageFactory.initElements(new AppiumFieldDecorator(driver), this);
` ?

Do you think that it's an issue I could report somewhere? Because I think that Page Factory is a popular solution in many test frameworks and lot's of people could encounter this problem when the XCUITest driver becomes available in some future Appium release.

Thanks again for responding.

Anna

**Comment by kaleposhobios** [comment]
Hi Anna,

It's a little beyond anything I've touched. I don't really know why WebDriverAgent doesn't support it. Maybe go to WebDriverAgent on GitHub and ask? 

I'm also not sure about the Selenium spec offhand, but it seems like in the Appium routing code they post directly to timeouts, in addition to timeout/set_implicit_timeout:

'/wd/hub/session/:sessionId/timeouts': {
    POST: {command: 'timeouts', payloadParams: {required: ['type', 'ms']}}
  },

So maybe that's the POST that's breaking WebDriverAgent. What if in the meantime you add a new route to the local WebDriverAgent code you already have (that appium-xcuitest-driver has as a submodule)?
I mean like:
WebDriverAgent/WebDriverAgentLib/Commands/FBCustomCommands.m
```
+ (NSArray *)routes
{
  return
  @[
    [[FBRoute POST:@"/deactivateApp"] respondWithTarget:self action:@selector(handleDeactivateAppCommand:)],
    [[FBRoute POST:@"/timeouts/implicit_wait"] respondWithTarget:self action:@selector(handleImplicitWaitCommand:)],
    [[FBRoute POST:@"/timeouts/"] respondWithTarget:self action:@selector(handleImplicitWaitCommand:)],
    [[FBRoute POST:@"/hide_keyboard"] respondWithTarget:self action:@selector(handleHideKeyboard:)]
  ];
}
```
(where ```/timeouts/``` is the new route to add).

I know that's not really an answer...the root of the question I think it beyond my knowledge of Appium and its intentions...

But I've just been using xcuitest-driver and WebDriverAgent a little. So, hopefully this might help at all.

--Matt

**Comment by arompa** [comment]
Hi Matt,

Thanks for all the help - I will try asking in Facebook WebdriverAgent about this :)

What you wrote does not seem to resolve my problem (maybe I did something wrong)

I will leave this Issue open - maybe someone who nows the new driver approach will have something to add.

Thanks,
Anna

**Comment by rafael-chavez** [comment]
do note, all the mobile commands in AppiumDriver.java (java-client) will also give you this unhandled exception. These commands need to be removed.

**Comment by kaleposhobios** [comment]
Do these all fall under the same route? If so, how about adding them to the proxy ```ignored``` list so they're not proxied, then in xcuidriver handling them more gracefully?

**Comment by rafael-chavez** [comment]
Most of these are non existent in this driver. For example:

"/session/:sessionId/appium/device/lock"

maps to command: 

    POST: {command: 'lock', payloadParams: {optional: ['seconds']}}

in the routes of MJWP. This is then executes the automation client (instruments) directly via:

uiAutoClient.sendCommand(`au.lock(${secs})`);

I think ideally we would like to see this functionality added to this new driver.

**Comment by kaleposhobios** [comment]
@rafael-chavez I have an outstanding pull request that addresses some of the missing features: https://github.com/appium/appium-xcuitest-driver/pull/25

Basically it adds an intercept and translation layer between the appium command, and the command WebDriverAgent is expecting.

I talked a little bit to @Jonahss and I understand there is probably a better way to handle this, I guess inside the proxy code itself. But that's beyond what I've touched.

Anyway doing something like that should let you do what you want, right? It's just dependent on whatever commands WebDriverAgent accepts. I've also submitted some PRs to WebDriverAgent and they've been great about accepting them.
